yb-voyager/cmd/importDataFileTaskPicker.go

```diff
@@ -409,3 +409,322 @@ func (c *ColocatedAwareRandomTaskPicker) WaitForTasksBatchesTobeImported() error
+		if err != nil {
+			return nil, fmt.Errorf("getting file import state for tasl: %v: %w", task, err)
+		}
+		switch taskStatus {
+		case FILE_IMPORT_COMPLETED:
+			doneTasks = append(doneTasks, task)
+		case FILE_IMPORT_IN_PROGRESS:
+			switch tableType {
+			case COLOCATED:
+				if len(inProgressColocatedTasks) < maxColocatedTasksInProgress {
+					addToInProgressTasks(task, tableType)
+				} else {
+					addToPendingTasks(task, tableType)
+				}
+			case SHARDED:
+				if len(inProgressShardedTasks) < maxShardedTasksInProgress {
+					addToInProgressTasks(task, tableType)
+				} else {
+					addToPendingTasks(task, tableType)
+				}
+			}
+		case FILE_IMPORT_NOT_STARTED:
+			addToPendingTasks(task, tableType)
+		default:
+			return nil, fmt.Errorf("unexpected  status for task: %v: %v", task, taskStatus)
+		}
+	}
+
+	picker := &ColocatedCappedRandomTaskPicker{
+		state:                       state,
+		maxShardedTasksInProgress:   maxShardedTasksInProgress,
+		maxColocatedTasksInProgress: maxColocatedTasksInProgress,
+
+		doneTasks: doneTasks,
+
+		inProgressColocatedTasks: inProgressColocatedTasks,
+		inProgressShardedTasks:   inProgressShardedTasks,
+
+		pendingColocatedTasks: pendingColcatedTasks,
+		pendingShardedTasks:   pendingShardedTasks,
+
+		tableTypes:              tableTypes,
+		colocatedBatchTaskQueue: colocatedBatchTaskQueue,
+	}
+
+	log.Infof("ColocatedCappedRandomTaskPicker initialized with params:%v", spew.Sdump(picker))
+	return picker, nil
+}
+
+func (c *ColocatedCappedRandomTaskPicker) inProgressTasks() []*ImportFileTask {
+	return append(c.inProgressColocatedTasks, c.inProgressShardedTasks...)
+}
+
+func (c *ColocatedCappedRandomTaskPicker) pendingTasks() []*ImportFileTask {
+	return append(c.pendingColocatedTasks, c.pendingShardedTasks...)
+}
+
+func (c *ColocatedCappedRandomTaskPicker) HasMoreTasks() bool {
+	return len(c.inProgressTasks()) > 0 || len(c.pendingTasks()) > 0
+}
+
+func (c *ColocatedCappedRandomTaskPicker) HasMoreColocatedTasks() bool {
+	return len(c.inProgressColocatedTasks) > 0 || len(c.pendingColocatedTasks) > 0
+}
+
+func (c *ColocatedCappedRandomTaskPicker) HasMoreShardedTasks() bool {
+	return len(c.inProgressShardedTasks) > 0 || len(c.pendingShardedTasks) > 0
+}
+
+func (c *ColocatedCappedRandomTaskPicker) pickRandomFromListOfTasks(tasks []*ImportFileTask) (int, *ImportFileTask) {
+	if len(tasks) == 0 {
+		panic("no tasks to pick from")
+	}
+	// pick a random task
+	taskIndex := rand.Intn(len(tasks))
+	return taskIndex, tasks[taskIndex]
+}
+
+func (c *ColocatedCappedRandomTaskPicker) Pick() (*ImportFileTask, error) {
+	if !c.HasMoreTasks() {
+		return nil, fmt.Errorf("no more tasks")
+	}
+
+	// if only one type of tasks are left, pick from that type.
+	if !c.HasMoreShardedTasks() {
+		return c.pickColocatedTask()
+	}
+	if !c.HasMoreColocatedTasks() {
+		return c.pickShardedTask()
+	}
+
+	// we have a combination of tasks left.
+	// first fill up in-progress tasks from pending tasks if possible.
```

Comment:
> // first fill up in-progress tasks from pending tasks if possible.

what is the reasoning behind this?

Reply:
In general, the point of the picker is to have multiple tables progressing at the same time, so the first thing that we do is fill up the maxTasksInProgress limit.